### ISSUE TITLE
Improve AI in combat not to waste units

### ISSUE THREAD

**Issue by fheroes2bugs** [ihhub/fheroes2#7789]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

I have noticed, that AI wants to win battle as fast as he can.
In regular situation (AI enemy has at least 1 castle), enemy hero surely will retreat, so for me, it is good idea to kill as much as I can (as fast as I can). So the logic is correct, I am ok with this.

But what about if enemy have no castles? He will not retreat, but my AI still wants to deal maximum damage in minumum turns, but these minimum are not actually the minimum. Logic "kill as much as I can (as fast as I can)" is not suitable for this situation.

Like in this video,  AI loses 2 units.

https://github.com/ihhub/fheroes2/assets/145407420/f82e1847-ce2d-4e79-bd63-ec5e94f35c15

But when I play it, I don't lose them:

https://github.com/ihhub/fheroes2/assets/145407420/5685b74b-5e14-4b0b-8b50-4112cc4d7fe4

[improve AI in combat.zip](https://github.com/ihhub/fheroes2/files/12694022/improve.AI.in.combat.zip)

### Describe the possible solution

1. If enemy have no castles, logic should not be "kill as much as I can (as fast as I can)". It should be "win, but with minumum loss of units".

2. In this video, when AI plays - turns of gydras and minotours are useless, they just waste our time. If playing AI vs AI and we are watching it - may be, it is better not to move them at all, just to not waste our time while we watch...

**Comment by Branikolog** [comment]
Hi, @fheroes2bugs 

>In this video, when AI plays - turns of gydras and minotours are useless, they just waste our time. If playing AI vs AI and we are watching it - may be, it is better not to move them at all, just to not waste our time while we watch...

There's always a reason to move all creatures:. What if opponent can demolish griffins with a powerful spell? What if some elementals would be summoned? Or low morale triggered for griffins?

It's just a machine. AI should move troops unless there's no possible movement, Even pro players tend to move forwards to be able to perform any action of something goes wrong.

>If enemy have no castles, logic should not be "kill as much as I can (as fast as I can)". It should be "win, but with minumum loss of units".

I think it's not the best reason for AI to use such tactics. Sometimes we should kill as fast, as we can. But sometimes not. It depends on the stage of the game, strength of both heroes, available magic spells, magic powers, intention to prevent opponent from fleeing, need for extra exp, etc

**Comment by fheroes2bugs** [comment]
> Hi, @fheroes2bugs
> 
> > In this video, when AI plays - turns of gydras and minotours are useless, they just waste our time. If playing AI vs AI and we are watching it - may be, it is better not to move them at all, just to not waste our time while we watch...
> 
> There's always a reason to move all creatures:. What if opponent can demolish griffins with a powerful spell? What if some elementals would be summoned? Or low morale triggered for griffins?
> 
> It's just a machine. AI should move troops unless there's no possible movement, Even pro players tend to move forwards to be able to perform any action of something goes wrong.
> 
> > If enemy have no castles, logic should not be "kill as much as I can (as fast as I can)". It should be "win, but with minumum loss of units".
> 
> I think it's not the best reason for AI to use such tactics. Sometimes we should kill as fast, as we can. But sometimes not. It depends on the stage of the game, strength of both heroes, available magic spells, magic powers, intention to prevent opponent from fleeing, need for extra exp, etc

That is correct, but in this situation AI loses units, and the real human is not, so here we can improve. 
Also I have maximum morality and luck, so low morale can't be triggered. And dragon in immune to spells, so he will kill enemy no matter what. Also if enemy would summon, I believe, dragon would kill that unit also, or I just used magic on it (I have like 40 spell power), so again, human is stronger than the AI.

**Comment by Branikolog** [comment]
@fheroes2bugs 
This behaviour is actually common for novice players, who tend to run and attack, but experience some minor losses. AI is just playing too straightforward here, as some humans do too. 

But nevertheless, I personally even can't invent the rules, how AI should behave in similar situations.
Maybe we need to implement some "safe" logic for AI, in a case there's no reason to hurry. But firstly, what are the conditions of haste in battle? As I mentioned above, there are so many variables. On default every player (human or AI) wants to win as fast, as he can. 
Apart from fighting neutral armies, more conditions should be described here too.

**Comment by ihhub** [comment]
I am closing this issue as we made a lot of changes in the battle AI since then.
